source/tutorials/file-sets.md

```diff
@@ -0,0 +1,661 @@
+Such path arguments are then [implicitly turned into sets](https://nixos.org/manual/nixpkgs/unstable/#sec-fileset-path-coercion) that contain _all_ files under the given path.
+In the previous trace this is indicated by `(all files in directory)`.
+
+:::{tip}
+The `trace` function pretty-prints its first argument and returns its second argument.
+But since you often just need the pretty-printing in `nix repl`, you can omit the second argument:
+
+```shell-session
+nix-repl> fs.trace ./.
+trace: /home/user (all files in directory)
+«lambda @ /nix/store/1czr278x24s3bl6qdnifpvm5z03wfi2p-nixpkgs-src/lib/fileset/default.nix:555:8»
+```
+:::
+
+Even though file sets conceptually contain local files, these files are *never* added to the Nix store unless explicitly requested.
+So You don't have to worry as much about accidentally copying secrets into the world-readable store.
+
+In this example, although we pretty-printed the home directory, no files were copied.
+This is in contrast to coercion of paths to strings such as in `"${./.}"`,
+which copies the whole directory to the Nix store on evaluation!
+
+:::{warning}
+When using [experimental Flakes](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake), local directories with a `flake.nix` are always copied into the Nix store *completely* unless it is a Git repository!
+:::
+
+This implicit coercion also works for files:
+
+```shell-session
+$ touch some-file
+```
+
+```shell-session
+nix-repl> fs.trace ./some-file
+trace: /home/user
+trace: - some-file (regular)
+```
+
+In addition to the included file, this also prints its [file type](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-readFileType).
+
+
+## Example project
+
+To further experiment with the library, make a sample project.
+Create a new directory, enter it,
+and set up `niv` to pin the Nixpkgs dependency:
+
+```shell-session
+$ mkdir fileset
+$ cd fileset
+$ nix-shell -p niv --run "niv init --nixpkgs nixos/nixpkgs --nixpkgs-branch master"
+```
+
+Then create a `default.nix` file with the following contents:
+
+```{code-block} nix
+:caption: default.nix
+{
+  system ? builtins.currentSystem,
+  sources ? import ./nix/sources.nix,
+}:
+let
+  pkgs = import sources.nixpkgs {
+    config = { };
+    overlays = [ ];
+    inherit system;
+  };
+in
+pkgs.callPackage ./build.nix { }
+```
+
+Add two source files to work with:
+
+```shell-session
+$ echo hello > hello.txt
+$ echo world > world.txt
+```
+
+## Adding files to the Nix store
+
+To add files in a given file set to the Nix store,
+we use [`toSource`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.toSource).
+This function needs a `root` attribute to know which path to use as the root of the resulting store path,
+but only exactly the files in the `fileset` argument are included in the result.
```

Comment:
"root of store path" is potentially confusing. Isn't that `/nix/store`?

There is no "but" here,  because one part doesn't contradict the other.

```suggestion
The argument to this function requires a `root` attribute to determine which source directory to copy to the store, and exactly the files in the `fileset` attribute are included in the result.
```

PS: I really don't like breaking sentences up into multiple lines, because edits are usually based on sentence, not phrase. Don't know about others, but dealing with multiple lines in the GitHub UI almost always make it more tedious for me.

Reply:
Used
> The argument to this function requires a `root` attribute to determine which source directory to copy to the store.
Only the files in the `fileset` attribute are included in the result.